In the REGEXP_REPLACE function, when in PostgreSQL compatibility mode, accept (but ignore/strip out) the 'g' flag.

Catchwa opened the pull request:
Addresses #1554 by removing the 'g' flag when in PostgreSQL mode, allowing the same query to run across both H2 and PostreSQL with the same behaviour.

Catchwa committed:
In the REGEXP_REPLACE function, when in PostgreSQL compatibility mode, accept (but ignore/strip out) the 'g' flag.

Catchwa committed:
Document change against the function definition

Catchwa commented:
Have added doco to the functions page (not sure where the best place to put it would be, but used the LOG function as a precedent).

katzyn commented on h2/src/main/org/h2/expression/Function.java:
```diff
@@ -1370,6 +1370,10 @@ protected Value getValueWithArgs(Session session, Expression[] args) {
                     v3.getString();
+            if(database.getMode().getEnum().equals(Mode.ModeEnum.PostgreSQL)) {
+                // PostgreSQL doesn't do global replaces without a 'g' flag, but H2 does
+                regexpMode = regexpMode.replaceAll("g", "");
```
From my point of view `makeRegexpFlags()` is a more efficient place to ignore the `g` flag. `String.replaceAll()` is not so fast.

Also you may want to improve PostgreSQL compatibility mode by introduction of conditional usage of `Matcher.replaceFirst()` instead of `Matcher.replaceAll()` a few lines below when H2 is in PostgreSQL compatibility mode and `regexpMode` does not contain the `g` character.

Catchwa replied:
Thanks - both great comments. Will amend.

katzyn commented on h2/src/docsrc/help/help.csv:
```diff
@@ -4296,7 +4296,7 @@ Replaces each substring that matches a regular expression.
 If any parameter is null (except optional flagsString parameter), the result is null.
 
-Flags values limited to 'i', 'c', 'n', 'm'. Other symbols causes exception.
+Flags values limited to 'i', 'c', 'n', 'm'. Other symbols causes exception (except for 'g' which is ignored when running in PostgreSQL mode).
```
Usually we don't describe such deviations in documentation of functions.

Here is a better place to describe it:
https://github.com/h2database/h2database/blob/e8f99400a3f71312f8edbc86f088becb67210cf5/h2/src/docsrc/html/features.html#L940-L957

Catchwa replied:
Thanks, will add something there.

katzyn reviewed (commented):
Please, add a some test to `regexp-replace.sql`.

See also comments below.

Catchwa committed:
Move documentation, add tests, refactor change logic

Add boolean flag to makeRegexpFlags that indicates whether the presence of a global flag should simply be ignored (i.e. when in PostreSQL mode) or whether it should throw an exception.

If the database is in PostgreSQL mode and the 'g' flag is not present, use the String.replaceFirst method instead of the H2 default String.replaceAll method when doing the regex replace.

Catchwa commented:
Not sure why the JDK10 build for my updates failed - doesn't seem to be related to my changes.

andreitokar commented:
Yes, it's probably unrelated. It had been failing a few times before. I will increase lock timeout for this test. Just ignore this failure for now.

katzyn commented on h2/src/docsrc/html/features.html:
```diff
@@ -950,7 +950,12 @@ <h3>PostgreSQL Compatibility Mode</h3>
+    <ul>
+    <li>Uses \ for back-references</li>
+    <li>Will not throw an exception when the <code>flagsString</code> parameter contains a 'g'</li>
+    <li>Will use Java's <code>String.replaceFirst()</code> method (instead of <code>String.replaceAll()</code>) if the 'g' flag is not present in the <code>flagsString</code> parameter.</li>
```
Implementation's details are not relevant here, may be just notice that in absence of 'g' flag only the first matched substring will be replaced?

katzyn commented on h2/src/main/org/h2/expression/Function.java:
```diff
@@ -1370,12 +1370,19 @@ protected Value getValueWithArgs(Session session, Expression[] args) {
             String regexpMode = v3 == null || v3.getString() == null ? "" :
                     v3.getString();
-            int flags = makeRegexpFlags(regexpMode);
+            int flags = makeRegexpFlags(regexpMode, Mode.ModeEnum.PostgreSQL.equals(database.getMode().getEnum()));
```
`Mode.ModeEnum.PostgreSQL.equals(database.getMode().getEnum())` can be extracted into local variable.

katzyn commented on h2/src/main/org/h2/expression/Function.java:
```diff
@@ -2049,6 +2056,10 @@ private static int makeRegexpFlags(String stringFlags) {
+                    case 'g':
+                        if (ignoreGlobalFlag) {
+                            break;
+                        }
```
Please, insert a new line with
```Java
//$FALL-THROUGH$
```
comment here, before the line with `default` operator.

Without it some compilers / IDEs show a warning.

katzyn reviewed (commented):
It looks mostly good, there are only a few minor suggestions below.

Please also send a license statement as described here:
http://h2database.com/html/build.html#providing_patches

It's better to send it to the mailing list:
https://groups.google.com/forum/#!forum/h2-database

Catchwa committed:
Create local variable for whether we're in PostgreSQL mode, remove implementation details from doco, add compiler/IDE hint for falling-through switch statement

Catchwa committed:
Add missing semicolon

Catchwa commented:
Have just posted to the H2 Google Group with the licensing statement.

katzyn commented:
Thank you!
